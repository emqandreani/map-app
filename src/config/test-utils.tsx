import { StyleSystemProvider } from "@architecture-it/stylesystem";
import { configureStore, PreloadedState } from "@reduxjs/toolkit";
import { render, renderHook } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import getTheme from "@architecture-it/stylesystem/Theme";
import React, { PropsWithChildren } from "react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";

import { ModuleContext } from "../store/hooks";
import { RootState, store } from "../store/store";
import { reducers } from "../store/reducers";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: typeof store;
  basename?: string;
}

interface WithProvidersProps {
  basename?: string;
  store: typeof store;
}

const colors = {
  primary: "#d71920",
  colorText: "#616161",
  error: "#d71920",
  secondary: "#b6040b",
};

export function WithProviders({ basename, store }: WithProvidersProps) {
  return function Wrapper({ children }: PropsWithChildren<{}>) {
    return (
      <BrowserRouter basename={basename}>
        <StyleSystemProvider theme={getTheme(false, colors)}>
          <Provider context={ModuleContext} store={store}>
            {children}
          </Provider>
        </StyleSystemProvider>
      </BrowserRouter>
    );
  };
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducers, preloadedState }),
    // Basename might be "/" for modules
    basename = "/",
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: WithProviders({ basename, store }), ...renderOptions }),
  };
}

export function renderHookWithProviders<Result, Props>(
  hook: (_initialProps: Props) => Result,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducers, preloadedState }),
    // Basename might be "/" for modules
    basename = "/",
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...renderHook(hook, { wrapper: WithProviders({ basename, store }), ...renderOptions }),
  };
}
