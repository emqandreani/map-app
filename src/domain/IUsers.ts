export interface IUsers {
  id: number;
  name: string;
  lastName: string;
  age: number;
}
