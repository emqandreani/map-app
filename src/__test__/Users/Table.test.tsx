import { renderWithProviders } from "../../config/test-utils";
import Table from "../../pages/Users/components/Table";
import { IUsers } from "../../domain/IUsers";

const header = ["Id", "Nombre", "Apellido", "Edad"];

const defaultData: IUsers[] = [
  {
    id: 1,
    name: "Lionel",
    lastName: "Messi",
    age: 36,
  },
];

describe("<Table />", () => {
  test("Should render without crush", () => {
    const { getByTestId } = renderWithProviders(Table({ header, data: defaultData }));

    const table = getByTestId("table");

    expect(table).toMatchSnapshot();
  });
  test("Table should render the correct header names", () => {
    const { getAllByTestId } = renderWithProviders(Table({ header, data: defaultData }));

    const headerCell = getAllByTestId("header-cell");

    headerCell.forEach((cell, index) => {
      expect(cell.textContent).toBe(header[index]);
    });
  });
  test("Table should render data exactly as passed", () => {
    const { getByText } = renderWithProviders(Table({ header, data: defaultData }));

    defaultData.forEach((item) => {
      expect(getByText(item.id)).toBeTruthy();
      expect(getByText(item.name)).toBeTruthy();
      expect(getByText(item.lastName)).toBeTruthy();
      expect(getByText(item.age.toString())).toBeTruthy();
    });
  });
});
