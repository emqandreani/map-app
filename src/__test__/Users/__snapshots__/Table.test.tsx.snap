// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> Should render without crush 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root tableContainer css-tt2mwc-MuiPaper-root-MuiTableContainer-root"
  data-testid="table"
>
  <table
    class="MuiTable-root css-1urm07n-MuiTable-root"
  >
    <thead
      class="MuiTableHead-root StyleSystem__Tablehead_head css-15wwp11-MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root StyleSystem__TableRow_tbody MuiTableRow-head css-p1m2vl-MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1824csp-MuiTableCell-root"
          data-testid="header-cell"
          scope="col"
        >
          Id
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1824csp-MuiTableCell-root"
          data-testid="header-cell"
          scope="col"
        >
          Nombre
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1824csp-MuiTableCell-root"
          data-testid="header-cell"
          scope="col"
        >
          Apellido
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1824csp-MuiTableCell-root"
          data-testid="header-cell"
          scope="col"
        >
          Edad
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root StyleSystem__TableRow_tbody css-p1m2vl-MuiTableRow-root"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1acrjjy-MuiTableCell-root"
        >
          1
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1acrjjy-MuiTableCell-root"
        >
          Lionel
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1acrjjy-MuiTableCell-root"
        >
          Messi
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1acrjjy-MuiTableCell-root"
        >
          36
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
